### PR TITLE
feat: add read apis for generic metrics

### PR DESCRIPTION
**Pull request by hamidzr**
## Description
Update the following apis to support arbitrary metric type(s) in a backward compatible way:
- MetricBatches
- SummarizeTrial
- CompareTrials
- ExpMetricNames
- GetTrainingMetrics
- GetValidationMetrics

Add
- GetMetrics

Refactor and bunify BatchMetrics

## Commentary (optional)

potential bug for metric_ids in multi trial sample we don't set metric type

do we want to only apply this to when metricIds are passed in?
// If no time series filter column name is supplied then default to batches.
same for maxDatapoints

### api changes

read:
- [separate pr] GetProjectNumericMetricsRange(
- [x] MetricBatches(
- [x] SummarizeTrial
- [x] CompareTrials
- [x] ExpMetricNames(
- [x] GetMetrics(
- [x] GetTrainingMetrics(
- [x] GetValidationMetrics(
write:
- [x] ReportTrialMetrics(
- [x] ReportTrialTrainingMetrics(
- [x] ReportTrialValidationMetrics(

some of the affected apis treated unspecified metric type as both training and validation or ALL metric types.
do we want to keep that at just those two or all, or error out? 

our latest API docs: https://docs.determined.ai/latest/rest-api/index.html

affected types:

1. MetricBatchesRequest
2. TrialsSnapshotRequest
3. TrialsSampleRequest
4. DownsampledMetrics
5. GetTrialWorkloadsRequest
6. SummarizeTrialRequest
7. CompareTrialsRequest

options: update we go with number 5

1. - change the metric type from enum to string on the existing endpoints (breaking change)
2. - change the metric type from enum to string on the existing endpoints but do some api and db level
3. cajoling? to make the affected APIs behave the same for training and validation metric types aka work with having the enum prefix. 
4. - set up duplicate (v2?) endpoints with just the metric type changed to string.
5. [x] Introduce new fields on the existing endpoints to add options for working with new metric types.

depends on https://github.com/determined-ai/determined/pull/7012

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

DET-9448
WEB-1346

### CODE REVIEW

File: master/internal/api_trials.go [hunk 1]
```diff
@@ -682,60 +698,32 @@ func (a *apiServer) MultiTrialSample(trialID int32, metricNames []string,
...
-			(metricType == apiv1.MetricType_METRIC_TYPE_UNSPECIFIED) {
-			var metric apiv1.DownsampledMetrics
-			metricMeasurements, err = trials.MetricsTimeSeries(
-				trialID, startTime, metricNames, startBatches, endBatches,
-				xAxisLabelMetrics, maxDatapoints, batches, timeSeriesFilter, "validation")
-			if err != nil {
-				return nil, errors.Wrapf(err, "error fetching time series of validation metrics")
-			}
-			metric.Type = apiv1.MetricType_METRIC_TYPE_VALIDATION
+	var timeSeriesColumn *string
```
Comment (hamidzr) [inline master/internal/api_trials.go:705]: there are some behavior changes here were these not made to be more compatible? (mentioned in the pr description)

File: master/internal/db/postgres_trial_metrics.go [hunk 2]
```diff
@@ -23,6 +24,19 @@ const (
 	GenericMetric MetricPartitionType = "GENERIC"
 )
 
+// BunSelectMetricsQuery sets up a bun select query for based on new metrics table
+// simplifying some weirdness we set up for pg10 support.
+func BunSelectMetricsQuery(metricType model.MetricType, inclArchived bool) *bun.SelectQuery {
```
Comment (hamidzr) [inline master/internal/db/postgres_trial_metrics.go:29]: inclArchived is incorrect. 
rename archivedEq *bool

File: master/internal/api_trials.go [hunk 3]
```diff
@@ -682,60 +698,32 @@ func (a *apiServer) MultiTrialSample(trialID int32, metricNames []string,
...
 	}
 
-	if len(metricIds) > 0 {
-		var timeSeriesColumn *string
-
-		// If no time series filter column name is supplied then default to batches.
-		defaultTimeSeriesColumn := batches
-		if timeSeriesFilter == nil || timeSeriesFilter.Name == nil {
-			timeSeriesColumn = &defaultTimeSeriesColumn
+	// TODO: we could throw out duplicates. (breaking change?)
```
Comment (ioga) [inline master/internal/api_trials.go:715]: I doubt having duplicates has ever been intentional. don't think this qualifies as "breaking"

File: master/internal/db/postgres_experiments.go [hunk 4]
```diff
@@ -126,55 +127,80 @@ func GetNonTerminalExperimentCount(ctx context.Context,
...
+	WITH all_metrics AS (
+		SELECT jsonb_object_keys(summary_metrics) as metric_type, jsonb_object_keys(summary_metrics->jsonb_object_keys(summary_metrics)) as metric_name
+		FROM trials
+		WHERE experiment_id IN (%s)
+	)
+	SELECT DISTINCT metric_type, metric_name
+	FROM all_metrics
+	ORDER BY metric_type, metric_name;
+	`, strings.Trim(strings.Join(strings.Fields(fmt.Sprint(experimentIDs)), ","), "[]"))
+	err := db.queryRows(query, &rows)
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/db/postgres_experiments.go:149]: can we do CTEs through Bun? I'll look again

[the remark this answers is not in the document]
Comment (ioga) [reply master/internal/db/postgres_experiments.go:149]: yes: https://bun.uptrace.dev/guide/query-common-table-expressions.html
also if you want to do a raw query, you can always do it with bun `NewRaw`.

[the remark this answers is not in the document]
Comment (ioga) [reply master/internal/db/postgres_experiments.go:149]: also, in this particular query you don't even need a CTE / `with`. This is just a nested query.

File: proto/src/determined/metric/v1/metric.proto [hunk 5]
```diff
@@ -0,0 +1,17 @@
+syntax = "proto3";
+
+package determined.metric.v1;
+option go_package = "github.com/determined-ai/determined/proto/pkg/metricv1";
+
+import "protoc-gen-swagger/options/annotations.proto";
+
+// MetricName packages metric name and type.
+message MetricName {
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply proto/src/determined/metric/v1/metric.proto:9]: hmm, this less than ideal input format for get api requests makes sense to me but why follow the same for our output when we could be more explicit?

[the remark this answers is not in the document]
Comment (hamidzr) [reply proto/src/determined/metric/v1/metric.proto:9]: also this reminds me we probably want some validation around inserting special characters in metric type.

[the remark this answers is not in the document]
Comment (ioga) [reply proto/src/determined/metric/v1/metric.proto:9]: if input and output format for the same entity has different types, one could say it's inconsistent.
e.g. if you request string metric id = "<TYPE>.<NAME>", but in the response it's an object `{type: type, name: name}`, it means you'll have to convert the object back to the string yourself to figure out where the response for what you've asked.

File: master/internal/db/postgres_experiments.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/db/postgres_experiments.go]: updated.

File: proto/src/determined/metric/v1/metric.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply proto/src/determined/metric/v1/metric.proto]: I'll leave this for another pr. it looks like missing tests and I'd like test more before I change it.

**Review by ioga — APPROVED** [review]

**Comment by hamidzr** [comment]
thanks for the reviews

**Pull request merged**